curated_transformers/pipe.py

```diff
@@ -280,28 +290,36 @@ def update(
         validate_examples(examples, "Transformer.update")
         docs = [eg.predicted for eg in examples]
         set_dropout_rate(self.model, drop)
-        tokvecs, bp_tokvecs = self.model.begin_update(docs)
-        d_tokvecs = [
-            Ragged(self.model.ops.alloc_f(t2v.dataXd.shape), t2v.lengths)
-            for t2v in tokvecs
-        ]
         losses.setdefault(self.name, 0.0)
 
+        if not self.frozen:
+            tokvecs, bp_tokvecs = self.model.begin_update(docs)
+            d_tokvecs = [
+                Ragged(self.model.ops.alloc_f(t2v.dataXd.shape), t2v.lengths)
+                for t2v in tokvecs
+            ]
+        else:
+            # Ensures that the inner torch model is executed in a `no_grad` context.
+            tokvecs = self.model.predict(docs)
+            bp_tokvecs = None
```

Comment:
My worry with the current implementation is: what if someone runs update when `frozen=True`, but then toggles `frozen=False` before the listeners backprop. This will then fail, because `d_tokvecs` is not allocated, etc.

I think `accumulate_gradient` and `backprop` below should be stateless (at least wrt. `self.frozen`). Instead, the listeners should receive a function/lambda that doesn't do anything when `self.frozen==True`. Then if someone switches to `frozen=False`, nothing bad happens when the listeners backpropagate.